File: apps/api/src/routes/roles.ts
/** Role management + permission catalog (SRS §6.7). Guarded by user.read / user.manage. */
import { Hono } from "hono";
import { HTTPException } from "hono/http-exception";
import {
  createRoleSchema,
  updateRoleSchema,
  PERMISSION_GROUPS,
  PERMISSION_KEYS,
} from "@tuition/shared";
import type { AppContext, Db } from "../types";
import { parseBody } from "../lib/validate";
import { newId, nowIso } from "../lib/id";
import { writeAudit } from "../lib/db";
import { authenticate, requirePermission } from "../middleware/auth";

const PERMISSION_KEY_SET = new Set(PERMISSION_KEYS);

/** Reject unknown permission keys; the `*` wildcard is reserved for the owner role. */
function validatePermissions(perms: string[]): void {
  for (const p of perms) {
    if (!PERMISSION_KEY_SET.has(p)) {
      throw new HTTPException(422, { message: "unknown_permission", cause: { permission: p } });
    }
  }
}

async function loadRole(db: Db, id: string) {
  const role = await db
    .prepare(`SELECT id, key, label, description, system FROM roles WHERE id = ?`)
    .bind(id)
    .first<{ id: string; key: string; label: string; description: string; system: number }>();
  if (!role) return null;
  const perms = await db
    .prepare(`SELECT permission_key FROM role_permissions WHERE role_id = ?`)
    .bind(id)
    .all<{ permission_key: string }>();
  return {
    id: role.id,
    key: role.key,
    label: role.label,
    description: role.description,
    system: role.system === 1,
    permissions: (perms.results ?? []).map((p) => p.permission_key),
  };
}

export const rolesRoutes = new Hono<AppContext>();
rolesRoutes.use("*", authenticate);

/** GET /api/roles — all roles with permissions and assigned-user counts. */
rolesRoutes.get("/", requirePermission("user.read"), async (c) => {
  const db = c.get("db");
  const roles = await db
    .prepare(`SELECT id, key, label, description, system FROM roles ORDER BY system DESC, label`)
    .all<{ id: string; key: string; label: string; description: string; system: number }>();
  const perms = await db
    .prepare(`SELECT role_id, permission_key FROM role_permissions`)
    .all<{ role_id: string; permission_key: string }>();
  const counts = await db
    .prepare(`SELECT role_id, COUNT(*) AS n FROM user_roles GROUP BY role_id`)
    .all<{ role_id: string; n: number }>();

  const permMap = new Map<string, string[]>();
  for (const p of perms.results ?? []) {
    const list = permMap.get(p.role_id) ?? [];
    list.push(p.permission_key);
    permMap.set(p.role_id, list);
  }
  const countMap = new Map((counts.results ?? []).map((r) => [r.role_id, r.n] as const));

  return c.json({
    roles: (roles.results ?? []).map((r) => ({
      id: r.id,
      key: r.key,
      label: r.label,
      description: r.description,
      system: r.system === 1,
      permissions: permMap.get(r.id) ?? [],
      user_count: countMap.get(r.id) ?? 0,
    })),
  });
});

/** POST /api/roles — create a custom role. */
rolesRoutes.post("/", requirePermission("user.manage"), async (c) => {
  const body = await parseBody(c, createRoleSchema);
  validatePermissions(body.permissions);
  const db = c.get("db");

  const clash = await db.prepare(`SELECT id FROM roles WHERE key = ?`).bind(body.key).first<{ id: string }>();
  if (clash) throw new HTTPException(409, { message: "role_key_taken" });

  const id = newId("rol");
  const now = nowIso();
  await db.batch([
    db
      .prepare(
        `INSERT INTO roles (id, key, label, description, system, created_at, updated_at)
         VALUES (?, ?, ?, ?, 0, ?, ?)`,
      )
      .bind(id, body.key, body.label, body.description, now, now),
    ...body.permissions.map((p) =>
      db.prepare(`INSERT INTO role_permissions (role_id, permission_key) VALUES (?, ?)`).bind(id, p),
    ),
  ]);
  await writeAudit(c.get("db"), {
    actorId: c.get("user").id,
    action: "role.create",
    entityType: "role",
    entityId: id,
    after: body,
  });
  return c.json(await loadRole(db, id), 201);
});

/** PATCH /api/roles/:id — edit label, description and/or permissions. */
rolesRoutes.patch("/:id", requirePermission("user.manage"), async (c) => {
  const body = await parseBody(c, updateRoleSchema);
  const db = c.get("db");
  const id = c.req.param("id");

  const role = await loadRole(db, id);
  if (!role) throw new HTTPException(404, { message: "not_found" });
  // The owner role must always retain the `*` wildcard — its permissions are locked.
  if (role.key === "owner" && body.permissions !== undefined) {
    throw new HTTPException(400, { message: "owner_permissions_locked" });
  }
  if (body.permissions !== undefined) validatePermissions(body.permissions);

  const stmts: D1PreparedStatement[] = [];
  const sets: string[] = [];
  const binds: unknown[] = [];
  if (body.label !== undefined) { sets.push("label = ?"); binds.push(body.label); }
  if (body.description !== undefined) { sets.push("description = ?"); binds.push(body.description); }
  if (sets.length > 0) {
    sets.push("updated_at = ?"); binds.push(nowIso());
    stmts.push(db.prepare(`UPDATE roles SET ${sets.join(", ")} WHERE id = ?`).bind(...binds, id));
  }
  if (body.permissions !== undefined) {
    stmts.push(db.prepare(`DELETE FROM role_permissions WHERE role_id = ?`).bind(id));
    for (const p of body.permissions) {
      stmts.push(db.prepare(`INSERT INTO role_permissions (role_id, permission_key) VALUES (?, ?)`).bind(id, p));
    }
  }
  if (stmts.length > 0) await db.batch(stmts);
  await writeAudit(db, { actorId: c.get("user").id, action: "role.update", entityType: "role", entityId: id, after: body });
  return c.json(await loadRole(db, id));
});

/** DELETE /api/roles/:id — delete a custom role (system roles and in-use roles are protected). */
rolesRoutes.delete("/:id", requirePermission("user.manage"), async (c) => {
  const db = c.get("db");
  const id = c.req.param("id");
  const role = await loadRole(db, id);
  if (!role) throw new HTTPException(404, { message: "not_found" });
  if (role.system) throw new HTTPException(400, { message: "cannot_delete_system_role" });

  const inUse = await db
    .prepare(`SELECT COUNT(*) AS n FROM user_roles WHERE role_id = ?`)
    .bind(id)
    .first<{ n: number }>();
  if ((inUse?.n ?? 0) > 0) throw new HTTPException(409, { message: "role_in_use" });

  await db.prepare(`DELETE FROM roles WHERE id = ?`).bind(id).run();
  await writeAudit(db, { actorId: c.get("user").id, action: "role.delete", entityType: "role", entityId: id });
  return c.json({ ok: true });
});

/** GET /api/permissions — the permission catalog, grouped for the matrix UI. */
export const permissionsRoutes = new Hono<AppContext>();
permissionsRoutes.use("*", authenticate);
permissionsRoutes.get("/", requirePermission("user.read"), (c) => {
  return c.json({ groups: PERMISSION_GROUPS });
});

File: apps/api/internal/roles/roles.go
// Package roles serves role management + the permission catalog
// (user.read / user.manage).
package roles

import (
	"context"
	"database/sql"
	"net/http"
	"strings"

	"github.com/go-chi/chi/v5"

	"attendly/api/internal/auth"
	"attendly/api/internal/cryptox"
	"attendly/api/internal/httpapi"
	"attendly/api/internal/rbac"
	"attendly/api/internal/store"
)

// Handlers serves the roles + permissions endpoints.
type Handlers struct{ db *sql.DB }

// New constructs the roles handlers.
func New(db *sql.DB) *Handlers { return &Handlers{db: db} }

// Mount registers /api/roles and /api/permissions (Authenticate upstream).
func (h *Handlers) Mount(r chi.Router) {
	read := auth.RequirePermission("user.read")
	manage := auth.RequirePermission("user.manage")
	r.Route("/api/roles", func(r chi.Router) {
		r.With(read).Method(http.MethodGet, "/", httpapi.Handler(h.list))
		r.With(manage).Method(http.MethodPost, "/", httpapi.Handler(h.create))
		r.With(manage).Method(http.MethodPatch, "/{id}", httpapi.Handler(h.update))
		r.With(manage).Method(http.MethodDelete, "/{id}", httpapi.Handler(h.remove))
	})
	r.With(read).Method(http.MethodGet, "/api/permissions", httpapi.Handler(h.permissions))
}

func (h *Handlers) actor(r *http.Request) string { return auth.MustUser(r.Context()).ID }

func validatePermissions(perms []string) error {
	for _, p := range perms {
		if !rbac.IsValidPermissionKey(p) {
			return httpapi.ErrWithDetails(http.StatusUnprocessableEntity, "unknown_permission", map[string]any{"permission": p})
		}
	}
	return nil
}

type roleDetail struct {
	ID          string   `json:"id"`
	Key         string   `json:"key"`
	Label       string   `json:"label"`
	Description string   `json:"description"`
	System      bool     `json:"system"`
	Permissions []string `json:"permissions"`
}

func (h *Handlers) loadRole(ctx context.Context, id string) (*roleDetail, error) {
	var rd roleDetail
	var system int
	err := h.db.QueryRowContext(ctx, `SELECT id, key, label, description, system FROM roles WHERE id = ?`, id).
		Scan(&rd.ID, &rd.Key, &rd.Label, &rd.Description, &system)
	if err == sql.ErrNoRows {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	rd.System = system == 1
	rd.Permissions = []string{}
	rows, err := h.db.QueryContext(ctx, `SELECT permission_key FROM role_permissions WHERE role_id = ?`, id)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	for rows.Next() {
		var p string
		if err := rows.Scan(&p); err != nil {
			return nil, err
		}
		rd.Permissions = append(rd.Permissions, p)
	}
	return &rd, rows.Err()
}

func (h *Handlers) list(w http.ResponseWriter, r *http.Request) error {
	roleRows, err := h.db.QueryContext(r.Context(), `SELECT id, key, label, description, system FROM roles ORDER BY system DESC, label`)
	if err != nil {
		return err
	}
	type roleView struct {
		roleDetail
		UserCount int64 `json:"user_count"`
	}
	roleList := []*roleView{}
	byID := map[string]*roleView{}
	for roleRows.Next() {
		var rv roleView
		var system int
		if err := roleRows.Scan(&rv.ID, &rv.Key, &rv.Label, &rv.Description, &system); err != nil {
			roleRows.Close()
			return err
		}
		rv.System = system == 1
		rv.Permissions = []string{}
		roleList = append(roleList, &rv)
		byID[rv.ID] = &rv
	}
	roleRows.Close()
	if err := roleRows.Err(); err != nil {
		return err
	}

	permRows, err := h.db.QueryContext(r.Context(), `SELECT role_id, permission_key FROM role_permissions`)
	if err != nil {
		return err
	}
	for permRows.Next() {
		var rid, pk string
		if err := permRows.Scan(&rid, &pk); err != nil {
			permRows.Close()
			return err
		}
		if rv := byID[rid]; rv != nil {
			rv.Permissions = append(rv.Permissions, pk)
		}
	}
	permRows.Close()

	countRows, err := h.db.QueryContext(r.Context(), `SELECT role_id, COUNT(*) FROM user_roles GROUP BY role_id`)
	if err != nil {
		return err
	}
	for countRows.Next() {
		var rid string
		var n int64
		if err := countRows.Scan(&rid, &n); err != nil {
			countRows.Close()
			return err
		}
		if rv := byID[rid]; rv != nil {
			rv.UserCount = n
		}
	}
	countRows.Close()

	httpapi.JSON(w, http.StatusOK, map[string]any{"roles": roleList})
	return nil
}

func (h *Handlers) create(w http.ResponseWriter, r *http.Request) error {
	var in struct {
		Key         string   `json:"key"`
		Label       string   `json:"label"`
		Description string   `json:"description"`
		Permissions []string `json:"permissions"`
	}
	if err := httpapi.Decode(r, &in); err != nil {
		return err
	}
	in.Key, in.Label = strings.TrimSpace(in.Key), strings.TrimSpace(in.Label)
	if in.Key == "" || in.Label == "" {
		return httpapi.BadRequest("invalid_body")
	}
	if err := validatePermissions(in.Permissions); err != nil {
		return err
	}
	var clash string
	if err := h.db.QueryRowContext(r.Context(), `SELECT id FROM roles WHERE key = ?`, in.Key).Scan(&clash); err == nil {
		return httpapi.Conflict("role_key_taken")
	} else if err != sql.ErrNoRows {
		return err
	}

	id := cryptox.NewID("rol")
	now := cryptox.NowISO()
	tx, err := h.db.BeginTx(r.Context(), nil)
	if err != nil {
		return err
	}
	defer func() { _ = tx.Rollback() }()
	if _, err := tx.ExecContext(r.Context(), `INSERT INTO roles (id, key, label, description, system, created_at, updated_at) VALUES (?, ?, ?, ?, 0, ?, ?)`, id, in.Key, in.Label, in.Description, now, now); err != nil {
		return err
	}
	for _, p := range in.Permissions {
		if _, err := tx.ExecContext(r.Context(), `INSERT INTO role_permissions (role_id, permission_key) VALUES (?, ?)`, id, p); err != nil {
			return err
		}
	}
	if err := tx.Commit(); err != nil {
		return err
	}
	actor := h.actor(r)
	_ = store.WriteAudit(r.Context(), h.db, store.AuditEntry{ActorID: &actor, Action: "role.create", EntityType: sp("role"), EntityID: &id})
	rd, err := h.loadRole(r.Context(), id)
	if err != nil {
		return err
	}
	httpapi.JSON(w, http.StatusCreated, rd)
	return nil
}

func (h *Handlers) update(w http.ResponseWriter, r *http.Request) error {
	id := chi.URLParam(r, "id")
	role, err := h.loadRole(r.Context(), id)
	if err != nil {
		return err
	}
	if role == nil {
		return httpapi.NotFound("not_found")
	}
	var in struct {
		Label       *string   `json:"label"`
		Description *string   `json:"description"`
		Permissions *[]string `json:"permissions"`
	}
	if err := httpapi.Decode(r, &in); err != nil {
		return err
	}
	if role.Key == "owner" && in.Permissions != nil {
		return httpapi.BadRequest("owner_permissions_locked")
	}
	if in.Permissions != nil {
		if err := validatePermissions(*in.Permissions); err != nil {
			return err
		}
	}

	tx, err := h.db.BeginTx(r.Context(), nil)
	if err != nil {
		return err
	}
	defer func() { _ = tx.Rollback() }()
	sets, args := []string{}, []any{}
	if in.Label != nil {
		sets, args = append(sets, "label = ?"), append(args, *in.Label)
	}
	if in.Description != nil {
		sets, args = append(sets, "description = ?"), append(args, *in.Description)
	}
	if len(sets) > 0 {
		sets = append(sets, "updated_at = ?")
		args = append(args, cryptox.NowISO())
		if _, err := tx.ExecContext(r.Context(), `UPDATE roles SET `+strings.Join(sets, ", ")+` WHERE id = ?`, append(args, id)...); err != nil {
			return err
		}
	}
	if in.Permissions != nil {
		if _, err := tx.ExecContext(r.Context(), `DELETE FROM role_permissions WHERE role_id = ?`, id); err != nil {
			return err
		}
		for _, p := range *in.Permissions {
			if _, err := tx.ExecContext(r.Context(), `INSERT INTO role_permissions (role_id, permission_key) VALUES (?, ?)`, id, p); err != nil {
				return err
			}
		}
	}
	if err := tx.Commit(); err != nil {
		return err
	}
	actor := h.actor(r)
	_ = store.WriteAudit(r.Context(), h.db, store.AuditEntry{ActorID: &actor, Action: "role.update", EntityType: sp("role"), EntityID: &id})
	rd, err := h.loadRole(r.Context(), id)
	if err != nil {
		return err
	}
	httpapi.JSON(w, http.StatusOK, rd)
	return nil
}

func (h *Handlers) remove(w http.ResponseWriter, r *http.Request) error {
	id := chi.URLParam(r, "id")
	role, err := h.loadRole(r.Context(), id)
	if err != nil {
		return err
	}
	if role == nil {
		return httpapi.NotFound("not_found")
	}
	if role.System {
		return httpapi.BadRequest("cannot_delete_system_role")
	}
	var inUse int64
	if err := h.db.QueryRowContext(r.Context(), `SELECT COUNT(*) FROM user_roles WHERE role_id = ?`, id).Scan(&inUse); err != nil {
		return err
	}
	if inUse > 0 {
		return httpapi.Conflict("role_in_use")
	}
	if _, err := h.db.ExecContext(r.Context(), `DELETE FROM roles WHERE id = ?`, id); err != nil {
		return err
	}
	actor := h.actor(r)
	_ = store.WriteAudit(r.Context(), h.db, store.AuditEntry{ActorID: &actor, Action: "role.delete", EntityType: sp("role"), EntityID: &id})
	httpapi.JSON(w, http.StatusOK, map[string]any{"ok": true})
	return nil
}

func (h *Handlers) permissions(w http.ResponseWriter, _ *http.Request) error {
	httpapi.JSON(w, http.StatusOK, map[string]any{"groups": rbac.PermissionGroups()})
	return nil
}

func sp(s string) *string { return &s }
